### ISSUE TITLE
ucp doesn't like using variables as source and destination

### ISSUE THREAD

**Issue by fdncred** [nushell/nushell#10300]
### Describe the bug

You cannot use variables for source and destination with ucp. See example below.

### How to reproduce

1. touch foo
2. mkdir test
3. let var = 'foo'
4. let folder = 'test'
5. cp -r $var $folder
```
❯ ucp -r $var $folder
Error:   × Missing file operand
   ╭─[entry #40:1:1]
 1 │ ucp -r $var $folder
   · ─┬─
   ·  ╰── Missing file operand
   ╰────
  help: Please provide source and destination paths
```

### Expected behavior

To expand variables

### Configuration

|key|value|
|-|-|
|version|0.84.1|
|branch|main|
|commit_hash|762fdb98acd7c22fcfb8a1fd82132884b483d566|
|build_os|macos-aarch64|
|build_target|aarch64-apple-darwin|
|rust_version|rustc 1.70.0 (90c541806 2023-05-31)|
|rust_channel|1.70.0-aarch64-apple-darwin|
|cargo_version|cargo 1.70.0 (ec8a8a0ca 2023-04-25)|
|build_time|2023-09-09 19:42:53 -05:00|
|build_rust_channel|release|
|allocator|mimalloc|
|features|dataframe, default, extra, sqlite, trash, which, zip|
|installed_plugins|emoji, file, from eml, from ics, from ini, from parquet, from plist, from vcf, gstat, highlight, hist, inc, json path, periodic-table, plot, pnet, query, query json, query web, query xml, regex, str bexpand, str similarity, xyplot|

### Additional context

/cc @dmatos2012 @tertsdiepraam

**Comment by tertsdiepraam** [comment]
Right, I guess my question wasn't clear. Let me try again. In most languages, it does not matter whether a function/command is passed a variable or a literal, because the runtime expands the variable automatically. Why does nushell differentiate between these two cases? In other words, why is this not supported automatically?

To be clear, I'm sure there's a good reason. I just want to understand why nushell works this way 😄

**Comment by fdncred** [comment]
I have no clue why. It seems like a logical thing to me.

**Comment by dmatos2012** [comment]
I think I found it, will be pushing a fix later, but Its because I filter the calls only by `Expr::FilePath` but `$var` is a `Expr::FullCellPath`, so therefore the paths is empty :( , see [here](https://github.com/nushell/nushell/blob/main/crates/nu-command/src/filesystem/ucp.rs#L19-L28)

**Comment by fdncred** [comment]
Thanks @dmatos2012. This bug has made me wonder, what if we have something like this:

```nushell
let files = [file1 file2 file3]
ucp $files /some/path
```
1. will your change support this?
2. should ucp/cp support lists like this?

I'm not sure what the right answer is to number 2.

**Comment by dmatos2012** [comment]
Hmm, good question, but theoretically, it wouldnt now, no because the signature is `SyntaxShape::FilePath`, so it would get rejected immediately by parser, but maybe we can change it to `SyntaxShapeOneOf?`? Not sure either what the answer is . My two cents is that we focus on stability like this sort of issues, and then we can see whether expanding makes sense

**Comment by tertsdiepraam** [comment]
> because I filter the calls

Maybe a filter is also not the best thing to do here? Instead of filtering I would expect an error to be returned if any of the arguments are wrong.

**Comment by tertsdiepraam** [comment]
I think I found the way to solve this. The variables should be evaluated using `Call` and the engine state and the stack. So instead of having our own `collect_filepath_arguments`, we should be using this function:
https://github.com/nushell/nushell/blob/762fdb98acd7c22fcfb8a1fd82132884b483d566/crates/nu-engine/src/call_ext.rs#L93

`open` also does this:
https://github.com/nushell/nushell/blob/762fdb98acd7c22fcfb8a1fd82132884b483d566/crates/nu-command/src/filesystem/open.rs#L66

Quick note as a newcomer to this code base: it was a bit difficult to figure out what all the types related to this were meant for. Some more documentation on `Expr`, `CallExt` & `FullCellPath` would be very useful, for example.

**Comment by dmatos2012** [comment]
Very nice observation, I didnt use it at first since we were dealing with lots of more options, but now that its been simplifed, I came to realize it was indeed better than rolling out my own. I had already fixed it that way couple of hours ago,  but had to stop to watch `TheEqualizer` lol, so now I am sitting down and writing a test for it.  Thanks for the help as well @tertsdiepraam.

```
![image](https://github.com/nushell/nushell/assets/11776554/77dd009b-42e3-46f3-ab8c-61fc29d6db4e)

```

**Comment by dmatos2012** [comment]
@tertsdiepraam any tips on knowing more of less which kind of things are supported or not on windows? I feel I am never sure if the tests will all pass on lets say windows, or how has `coreutils` managed to do that, and know in advance?. On latest fix, now suddenly it fails on [windows](https://github.com/nushell/nushell/actions/runs/6139123976/job/16656704338?pr=10304), and it looks two of them were on the ones where I changed the test, but one completely unrelated (copy_to_non_existing_dir).

Doing `#[target(not(windows))` will solve it but doesnt seems like the correct solution here!

**Comment by tertsdiepraam** [comment]
@dmatos2012 Yeah there's not really a good way of doing it sadly. I'd say, try a run on the CI first with plenty of `dbg!` and the remove those once it passes. In uutils, we log every util run to stderr as well I think. That makes it a bit easier to see what's going on.

**Comment by fdncred** [comment]
Darn, it just looks like Windows tests failing. I hate that. I definitely don't want to comment out tests on Windows if we don't have to. I'm afraid this iterative approach may be best.

If it's just the error text that is different, you can always have an expected windows error message and an expected non_windows error message.

**Comment by dmatos2012** [comment]
Yeah, I just realized I had some hardcoded `/` paths in there which of course doesnt work on Windows. I am sure there are more bugs to come, but is nice that this one is "fixed" with the latest push on the PR #10304
